core-bundle/src/DependencyInjection/Compiler/LoggerChannelPass.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\DependencyInjection\Compiler;
+
+use Contao\CoreBundle\Monolog\SystemLogger;
+use Symfony\Component\DependencyInjection\ChildDefinition;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+
+/**
+ * @internal
+ */
+class LoggerChannelPass implements CompilerPassInterface
+{
+    public const LEGACY_ACTIONS = [
+        'error',
+        'access',
+        'general',
+        'files',
+        'cron',
+        'forms',
+        'email',
+        'configuration',
+        'newsletter',
+    ];
+
+    private array $loggers = [];
+
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('monolog.logger')) {
+            return;
+        }
+
+        if ($container->hasParameter('contao.monolog.default_channels')) {
+            foreach ($container->getParameter('contao.monolog.default_channels') as $action) {
+                $id = 'monolog.logger.contao.'.$action;
+
+                $logger = clone $container->getDefinition('monolog.logger');
+                $logger->replaceArgument(0, 'contao.'.$action);
+
+                $container->setDefinition(
+                    $id,
+                    (new Definition(SystemLogger::class, [$logger, $this->transformContaoAction($action)]))
+                        // Public service for legacy use without dependency injection
+                        ->setPublic(true)
+                );
+
+                $this->loggers[] = $id;
+            }
+        }
+
+        foreach ($container->getDefinitions() as $id => $definition) {
+            if (!\in_array($id, $this->loggers, true) && $this->isContaoChannelLoggerDefinition($definition)) {
+                $container->setDefinition(
+                    $id,
+                    new Definition(SystemLogger::class, [$definition, $this->getContaoActionFromChannel($definition->getArgument(0))])
+                );
```

Comment:
This should create a decorated service instead of overriding the monolog service. Something like that:

```php
$container
    ->register("contao._logger.$id", SystemLogger::class)
    ->setDecoratedService($id)
    ->setArguments([new Reference("contao._logger.$id.inner"), $this->getContaoActionFromChannel($definition->getArgument(0))]);
```

We could even have a template service (similar to `monolog.logger_prototype`) but that might be over the tops.

Reply:
You are right. I have changed this in https://github.com/contao/contao/pull/3702/commits/8f6649e0b6510d78b42086fc072642a624080dd9 – simpler and more transparent.